File: src/days/day4.ts
import { readFile } from "../utils/index.js";

async function day4(): Promise<void> {
    const input = await readFile("day4.txt");
    const split = input.split("\n").map((x) => x.split(",")).map((x) => x.map((x) => x.split("-")));
    const filled = split.map((x) => {
        return x.map((v) => {
            const [start, end] = v;
            const range = Array.from(Array(parseInt(end) - parseInt(start) + 1).keys()).map((x) => x + parseInt(start));
            return range;
        })
    })
    const part1 = []
    const part2 = [];
    for (let i in filled) {
        const [first, second] = filled[i];
        if (first[0] <= second[0] && first[first.length - 1] >= second[second.length - 1]) {
            part1.push(1)
            part2.push(1);
        } else if (second[0] <= first[0] && second[second.length - 1] >= first[first.length - 1]) {
            part1.push(1)
            part2.push(1)
        } else {
            if (first[0] <= second[0] && first[first.length - 1] >= second[0]) {
                part2.push(1);
            }
            if (second[0] <= first[0] && second[second.length - 1] >= first[0]) {
                part2.push(1);
            }
            part1.push(0)
            part2.push(0)
        }
    }
    console.log(part1.reduce((a,b) => a+b), part2.reduce((a, b) => a + b));
}

await day4()

File: python/day4.py
lines = open('../inputs/day4.txt', 'r').read().split("\n")

lines = [[[int(v) for v in x.split("-")] for x in line.split(',')] for line in lines]
part1 = []
part2 = []
for line in lines:
    first = [x for x in range(line[0][0], line[0][1] + 1)]
    second = [x for x in range(line[1][0], line[1][1] + 1)]
    if first[0] <= second[0] and first[len(first) - 1] >= second[len(second) - 1]:
        part2.append(1)
        part1.append(1)
    elif second[0] <= first[0] and second[len(second) - 1] >= first[len(first) - 1]:
        part2.append(1)
        part1.append(1)
    else:
        if first[0] <= second[0] and first[len(first) - 1] >= second[0]:
            part2.append(1)
        elif second[0] <= first[0] and second[len(second) - 1] >= first[0]:
            part2.append(1)
        else:
            part2.append(0)
            part1.append(0)

print(sum(part2), sum(part1))